dash-pipeline/SAI/src/p4meta.h

```diff
@@ -0,0 +1,439 @@
+                std::initializer_list<std::map<uint32_t, P4MetaAction>::value_type> init_actions
+                ) : id(table_id), keys(init_keys), actions(init_actions)
+
+        {}
+
+        const P4MetaKey* get_meta_key(
+                _In_ sai_attr_id_t attr_id) const
+        {
+            for (auto i=0u; i<keys.size(); i++) {
+                if (attr_id == keys[i].attr_id) {
+                    return &keys[i];
+                }
+            }
+
+            return nullptr;
+        }
+
+        const P4MetaKey* get_meta_key(
+                _In_ const std::string &key_name) const
+        {
+            for (auto i=0u; i<keys.size(); i++) {
+                if (key_name == keys[i].name) {
+                    return &keys[i];
+                }
+            }
+
+            return nullptr;
+        }
+
+        const P4MetaActionParam* get_meta_action_param(
+                _In_ uint32_t action_id,
+                _In_ sai_attr_id_t attr_id) const
+        {
+            auto itr = actions.find(action_id);
+            if (itr != actions.end()) {
+                auto params = itr->second.params;
+                for (auto i=0u; i<params.size(); i++) {
+                    if (attr_id == params[i].attr_id) {
+                        return &params[i];
+                    }
+                }
+            }
+
+            return nullptr;
+        }
+    };
+
+    inline bool has_suffix(const std::string &str, const std::string &suffix)
+    {
+        if (str.length() < suffix.length())
+            return false;
+
+        return !str.compare(str.length() - suffix.length(), suffix.length(), suffix);
+    }
+
+    template<typename T, typename V>
+    static inline
+    void  set_value(
+            _In_ const std::string &field,
+            _In_ uint32_t bitwidth,
+            _In_ const V &value,
+            _Inout_ T *t)
+    {
+        if (field == "booldata")
+            return booldataSetVal(value, t, bitwidth);
+        if (field == "u8")
+            return u8SetVal(value, t, bitwidth);
+        if (field == "u16")
+            return u16SetVal(value, t, bitwidth);
+        if (field == "u32")
+            return u32SetVal(value, t, bitwidth);
+        if (field == "u64")
+            return u64SetVal(value, t, bitwidth);
+        if (field == "ipaddr")
+            return ipaddrSetVal(value, t, bitwidth);
+        if (field == "mac")
+            return macSetVal(value, t, bitwidth);
+        if (field == "u8list")
+            return u8listSetVal(value, t, bitwidth);
+
+        assert(0);
+    }
+
+    template<typename V>
+    static inline
+    void  set_value(
+            _In_ const std::string &field,
+            _In_ uint32_t bitwidth,
+            _In_ const V &value,
+            _Inout_ p4::v1::FieldMatch_LPM *t)
+    {
+        assert (field == "ipPrefix");
+        return ipPrefixSetVal(value, t, bitwidth);
+    }
+
+    template<typename T>
+    static inline
+    void get_value(
```

Comment:
better rename the functions. the names are too vague.

Reply:
fixed. Update functions to have a meaningful name.